### PR TITLE
[codex] Unify 2CR reproduction score comparison

### LINKED ISSUE #2675: Centralize 2CR score comparison and unify OK/OKish/FAIL rules
## Problem

Pyserini’s 2CR runners independently classify observed scores as `OK`, `OKish`, or `FAIL`. `_base.py` centralizes the display labels, but comparison logic is duplicated across modules, with different thresholds and treatment of improvements.

## Desired behavior

Introduce a shared classification helper in `pyserini/2cr/_base.py`, following the definitions proposed in [Anserini #3425](https://github.com/castorini/anserini/issues/3425), with configured tolerance treated as zero.

Let `delta = abs(observed - expected)`. Apply these checks **in order**:

1. **OK:** `delta ≤ numericalTolerance`, where `numericalTolerance` is a small, explicitly defined shared floating-point allowance.
2. **OKish:** If not already `OK`, either:
   - `observed > expected`
   - `delta < 0.0002`
3. **FAIL:** Otherwise.

Remove all existing runner-specific hardcoded tolerances, including MS MARCO’s `flaky_dict`. Do not introduce YAML tolerance configuration or benchmark-specific exceptions.

The numerical allowance and strict `< 0.0002` fallback remain shared policy constants. Normalize percentage-scale metrics to a 0–1 scale for comparison, while preserving existing score presentation.

The helper should return a shared status representation; runners remain responsible for evaluation, configuration loading, and output formatting.

## Behavior changes

- Regressions previously accepted under looser thresholds may now fail.
- Improvements outside the numerical allowance consistently become `OKish`.
- CIRAL and ODQA adopt the same three-status policy.
- Existing special-case tolerance allowances are removed.

## Acceptance criteria

- All 2CR runners delegate classification to the shared helper.
- Runner-specific comparison thresholds and `flaky_dict` are removed.
- The numerical tolerance is explicitly defined and documented.
- Comparison uses consistent score units while preserving existing display formats and evaluation precision.
- Tests cover equality, floating-point differences, improvements, regressions, the strict `0.0002` boundary, and equivalent fractional/percentage inputs.

### PR DESCRIPTION
**Pull request by lintool**
## Summary

- Centralize all 11 runners on `compare_reproduction_score(observed, expected)`, returning `ReproductionStatus`: differences ≤ `1e-9` are OK; otherwise improvements or differences strictly below `0.0002` are OKish; remaining results fail.
- Compare decimal string representations to avoid binary subtraction artifacts at the thresholds, without fixed-decimal rounding or additional tolerance.
- Remove runner-specific thresholds and MS MARCO’s `flaky_dict`. DSE and ODQA normalize percentage scores before comparison.
- Share status formatting through `format_reproduction_status`, preserving evaluation precision and score presentation.
- Add tests for equality, numerical differences, improvements, regressions, nonzero boundaries, normalized percentages, and all runners.

Closes #2675.

## Validation

- `mamba run -n pyserini-dev3 python -m unittest tests.base.test_2cr -v`
- `git diff --check`

All 7 tests pass, including 55 runner scenarios with mocked evaluation scores. Full retrieval benchmarks were not run.

### CODE REVIEW

File: pyserini/2cr/_base.py [hunk 1]
```diff
@@ -23,6 +24,38 @@
...
+    results fail. There is no relative or configurable tolerance.
+    """
+    if percentage:
+        observed /= 100
+        expected /= 100
+
+    delta = abs(observed - expected)
+    if delta <= NUMERICAL_TOLERANCE:
+        return ScoreStatus.OK
+    if observed > expected or delta < OKISH_THRESHOLD:
```
[the remark this answers is not in the document]
Comment (lintool) [reply pyserini/2cr/_base.py]: Fixed in 986b7f1cf. The helper now compares decimal string representations of the scores and thresholds, so `0.5000` versus `0.5002` produces FAIL without fixed-decimal rounding or additional tolerance.

Percentage normalization now lives in the DSE and ODQA callers; the helper takes only observed and expected scores. Tests cover normalized `50.00` versus `50.02`, nonzero boundary cases, and values immediately below and above the cutoff. All 7 tests pass, including 55 runner scenarios with mocked evaluation scores.

File: pyserini/2cr/odqa.py [hunk 2]
```diff
@@ -430,10 +429,8 @@ def run_topic_conditions(args, topic_arg, default_topics, yaml_path):
                     if not args.skip_eval and metric not in score.keys():
                         continue
                     if not args.skip_eval:
-                        if math.isclose(score[metric], float(expected_score), abs_tol=2e-2):
-                            result_str = ok_str
-                        else:
-                            result_str = fail_str + f' expected {expected[metric]:.4f}'
+                        status = compare_reproduction_score(score[metric] / 100, float(expected_score) / 100)
```
[the remark this answers is not in the document]
Comment (lintool) [reply pyserini/2cr/odqa.py:432]: Will punt on this because we'll refactor this later.

File: pyserini/2cr/atomic.py [hunk 3]
```diff
@@ -224,13 +222,8 @@ def run_conditions(args):
...
 
-                            if math.isclose(score, float(expected[metric])):
-                                result = ok_str
-                            # If results are within 0.0005, just call it "OKish".
-                            elif abs(score - float(expected[metric])) <= 0.0005:
-                                result = okish_str + f' expected {expected[metric]:.4f}'
-                            else:
-                                result = fail_str + f' expected {expected[metric]:.4f}'
+                            status = compare_reproduction_score(score, float(expected[metric]))
+                            result = format_reproduction_status(status, expected[metric])
```
Comment (lilyjge) [inline pyserini/2cr/atomic.py:226]: nit but why is one casted to float and one isn't? in the method signatures both methods are expecting floats

File: pyserini/2cr/msmarco.py [hunk 4]
```diff
@@ -503,16 +507,6 @@ def generate_report(args):
             out.write(Template(html_template).substitute(title=full_name, rows=all_rows))
 
 
-FlakyKey = namedtuple('FlakyKey', ['collection', 'name', 'topic_key', 'metric'])
```
Comment (lilyjge) [inline pyserini/2cr/msmarco.py:506]: just double checking that this removal is intended

**Review by lilyjge — APPROVED** [review]
the number of places that this needed to be changed makes me think that some abstraction might be useful, but that can be a later problem. nits but otherwise lgtm!

File: pyserini/2cr/msmarco.py (hunk 4 above)
Comment (lintool) [reply pyserini/2cr/msmarco.py:506]: correct.

File: pyserini/2cr/atomic.py (hunk 3 above)
Comment (lintool) [reply pyserini/2cr/atomic.py]: Good catch—the inconsistent cast was carried over from the original comparison and formatting code. Fixed in 22908dbb9: each runner now converts the expected score to float once and reuses it for both helpers. DSE and ODQA still divide by 100 for comparison and retain percentage units for display. All 7 tests pass, including 55 runner scenarios.

File: pyserini/2cr/dse.py [hunk 5]
```diff
@@ -125,13 +129,9 @@ def run_conditions(args):
...
 
-                            if math.isclose(score, float(expected[metric]), abs_tol=0.1): 
-                                result = ok_str
-                            elif abs(score - float(expected[metric])) <= 0.5:
-                                result = okish_str + f' expected {expected[metric]:.1f}'
-                            else:
-                                result = fail_str + f' expected {expected[metric]:.1f}'
-                            
+                            expected_score = float(expected[metric])
+                            status = compare_reproduction_score(score / 100, expected_score / 100)
```
[the remark this answers is not in the document]
Comment (lintool) [reply pyserini/2cr/dse.py:133]: we'll pick up the DSE repro refactoring later.
